### ISSUE TITLE
add inet_ntop() call: IPv6 support

### ISSUE THREAD

**Issue by brendangregg** [iovisor/bpftrace#30]
This will be useful for all networking one-liners/scripts. See the man page for inet_ntop(3), but maybe we should just call it `ntop()` for short (or something else).

Both of these should work, and print "127.0.0.1" and "::1":

```
bpftrace -e 'BEGIN { print("%s\n", ntop(AF_INET, 1)); }'
bpftrace -e 'BEGIN { print("%s\n", ntop(AF_INET6, 1)); }'
```

IPv4 should be shown as a dotted-quad, and IPV6 as a most-shortened address representation:   https://tools.ietf.org/html/rfc5952

**Comment by dalehamel** [comment]
With https://github.com/iovisor/bpftrace/pull/262 merged, I now have a practical reason to implement this support so I'm going to take a crack at it. My early attempts are crashing, I suspect a bug in codegen. Down the rabbit hole I go :rabbit: .

I noticed that bpftrace supports bitwise operations, but not hex literals (as far as I can tell). A quick and dirty way to translate the existing (now parsable) struct members such as `sk->__sk_common.skc_rcv_addr` which contain AF_INET addresses would be to just leverage bit-shifting to extract each bit of at least at ipv4 address for printing. As a work-around, I'm just printing hex values for now, though ports are thankfully simple integers and are easier to work with.

The implementation specified above would also probably need to specify the literal decimal `2` instead of `AF_INET`, as the current header parsing doesn't parse `#define` macros (though I suspect the support is there in clang to do so). I suspect that based on the buffer size, we could just infer whether an address is ipv4 or ipv6 though, so perhaps specifying the address type is an unnecessary part of the API.

**Comment by dalehamel** [comment]
I noticed in https://github.com/iovisor/bpftrace/issues/59 that hex literals are permitted, I guess they just have to be declared as a variable explicitly.

I was able to put together a way to print ipv4 addresses, but uses up 4 printf fields, so it'll mean using multiple printfs per line, but it'll get the job done in the meantime as a reasonable workaround:

```
  $sk->__sk_common.skc_rcv_daddr
  $mask1 = 0x000000FF;
  $mask2 = 0x0000FF00;
  $mask3 = 0x00FF0000;
  $mask4 = 0xFF000000;

  $q1 = ($daddr & $mask1) >> 0;
  $q2 = ($daddr & $mask2) >> 8;
  $q3 = ($daddr & $mask3) >> 16;
  $q4 = ($daddr & $mask4) >> 24;
  printf("%d.%d.%d.%d\n", $q1, $q2, $q3, $q4);
```

Sample output:

```
bpftrace -e 'BEGIN { $daddr = 0x08B2940A; $mask1 = 0x000000FF;  $mask2 = 0x0000FF00; $mask3 = 0x00FF0000; $mask4 = 0xFF000000; $q1 = ($daddr & $mask1) >> 0; $q2 = ($daddr & $mask2) >> 8; $q3 = ($daddr & $mask3) >> 16; $q4 = ($daddr & $mask4) >> 24; printf("%d.%d.%d.%d\n", $q1, $q2, $q3, $q4);}'
Attaching 1 probe...
10.148.178.8
```

**Comment by ajor** [comment]
Hex literals are meant to be fully supported in bpftrace - this program works for me without first assigning to variables:
`BEGIN { printf("%x\n", 0x4567 & 0xff) }`

**Comment by dalehamel** [comment]
@ajor hmm yes you seem to be right. I was getting a syntax error earlier but it might have due to the syntax I used for the bitwise operators, and when I switched to variables it was fine. Perhaps I was on an old branch of bpftrace?

Either way, yes you're correct, hex literals seem to be working just fine.

**Comment by dalehamel** [comment]
Accessing ipv6 addresses in the kernel is going to be a bit tricky https://github.com/iovisor/bpftrace/issues/268#issuecomment-441787339

**Comment by dalehamel** [comment]
I've completed an initial implementation of this in my PR #269, which I think is now ready for review.

**Comment by dalehamel** [comment]
In #275 I indicate a possible approach for handling array types, which is the issue with ipv6 addresses right now.

This would also allow for accepting `Type::array` as the second field to the `ntop` function in the semantic analyzer, which could then just copy the bytes down to `resolve_inet` directly, and I think is the cleanest way to get ipv6 support.

**Comment by brendangregg** [comment]
About IPv6: passing a 128 bit integer is one way to do it, but did you try passing it as 2 x 64 bit integers? I did such packing with usym(), for example:

```
  else if (call.func == "usym")
  {
    // store uint64_t[2] with: [0]: (uint64_t)addr, [1]: (uint64_t)pid
    AllocaInst *buf = b_.CreateAllocaBPF(call.type, "usym");
    b_.CreateMemSet(buf, b_.getInt8(0), call.type.size, 1);
    Value *pid = b_.CreateLShr(b_.CreateGetPidTgid(), 32);
    Value *addr_offset = b_.CreateGEP(buf, b_.getInt64(0));
    Value *pid_offset = b_.CreateGEP(buf, {b_.getInt64(0), b_.getInt64(8)});
    call.vargs->front()->accept(*this);
    b_.CreateStore(expr_, addr_offset);
    b_.CreateStore(pid, pid_offset);
    expr_ = buf;
  }
```

**Comment by brendangregg** [comment]
Changed title to adding IPv6 support.

Was just talking to @jasonk000 who might try coding it.

User space needs to take the bytes and format it. I did the code for bcc: https://github.com/iovisor/bcc/commit/9e0b087fceda18641eadbd567d3823c69d5f89a8

I think there's at least two approaches:

- A) pass by value: this probably needs #284, so we can have a 128-bit integer as the argument.
- B) pass by reference: this probably needs #275, so merging of #285.

Here's an example address (needs Linux 4.16+ for this tracepoint):

```
# bpftrace -lv t:sock:inet_sock_set_state
tracepoint:sock:inet_sock_set_state
    const void * skaddr;
    int oldstate;
    int newstate;
    __u16 sport;
    __u16 dport;
    __u16 family;
    __u8 protocol;
    __u8 saddr[4];
    __u8 daddr[4];
    __u8 saddr_v6[16];
    __u8 daddr_v6[16];
```

Hmmm. How would the code to either (A) or (B) look? I'm not sure. Maybe for (A):

```
bpftrace -e 'BEGIN { print("%s\n", ntop(AF_INET6, args->saddr_v6[0])); }'
# or
bpftrace -e 'BEGIN { print("%s\n", ntop(AF_INET6, (u128)args->saddr_v6[0])); }'
```

The first may not work since ntop() may only see the u8 and not the address. The second might be a way to cast it as a 128-bit int, but we'd need int casts first (#554) and it just may not make sense with the parser anyway -- since it may end up casing the u8 as a u128 and not getting the rest of the bits.

Hence (B):

```
bpftrace -e 'BEGIN { print("%s\n", ntop(AF_INET6, &args->saddr_v6[0])); }'
```

This will need #553. It probably makes the most sense.

It begs the question: do we change the ntop() API? For AF_INET, do we switch that to pass-by-reference for consistency, or leave it as pass-by-value since it's a 32-bit? I don't feel strongly. Consistency sounds good, if it works.

Fixing should fix #521 as a side effect.

**Comment by mmarchini** [comment]
While working on #553 I ended up implementing this (sort of) and #521. The only problem is, inet_ntop doesn't seem to support IPv6 yet:

```
$ bpftrace -e 'tracepoint:tcp:tcp_set_state { printf("%s\n", ntop(10, args->saddr_v6))  }'
Attaching 1 probe...
ntop() currently only supports AF_INET (IPv4); IPv6 will be supported in the future.

ntop() currently only supports AF_INET (IPv4); IPv6 will be supported in the future.

ntop() currently only supports AF_INET (IPv4); IPv6 will be supported in the future.

ntop() currently only supports AF_INET (IPv4); IPv6 will be supported in the future.

ntop() currently only supports AF_INET (IPv4); IPv6 will be supported in the future.

```

But maybe it's my Kernel version? I'll try with a newer kernel soon.

**Comment by dalehamel** [comment]
@mmarchini that message is internal to bpftrace https://github.com/iovisor/bpftrace/blob/48619f43e06c59c4454f80c114ddf3f97e8a78da/src/bpftrace.cpp#L1618-L1631

it is hardcoded to not work with the IPV6 addr family as there is currently no way to pass 128 bit types by value (or at least, there wasn't when we added the ipv4 functionality)

What you are passing there I think is stored as a constant array, and we need to either:

* Cast it to a 128 bit type to pass to ntop (there is already space allocated for this in the call, we just have no way to load it). This is the "by-value" approach @brendangregg mentions
* Get the address of the constant array, and store it in its own variable to pass by reference to ntop, changing the ntop api. This is the "by-reference" approach.

**Comment by mmarchini** [comment]
> Cast it to a 128 bit type to pass to ntop (there is already space allocated for this in the call, we just have no way to load it). This is the "by-value" approach @brendangregg mentions

I went essentially with this approach. https://github.com/iovisor/bpftrace/pull/566 should fix this and #553.

**Comment by dalehamel** [comment]
Awesome!!! Reading through it now, it looks great on first glance
